### PR TITLE
[DNM] [RFC] drivers: i2c: Add i2c_async for asynchronous transfers

### PR DESCRIPTION
**Pull request by nordic-krch**
### PR content:

- Proposed extension to i2c API:
  - It's is mainly addition of `i2c_single_transfer(dev, msg, add, callback, user_data)`.
  - ~~Extending `i2c_msg` flags to contain milliseconds delay field which can be used within sequence to delay message execution.~~
  - ~~Added helper macros for initializing set of message, e.g. `I2C_MSG_TXRX()`~~
  - Added function for getting `i2c_async` instance associated with given device: `i2c_get_async(dev)`
   
- Added new module `drivers/i2c/i2c_async` which wraps `queue_operation_manager` and `async_seq_manager` to provide safe and managed, asynchronous access to i2c.
  - Module is a fairly simple wrapper which binds those two modules together and finally uses `i2c_single_transfer()` to execute item within the sequence.
  - I2C sequence may consist of many `i2c_async_msg` items and may be constant, e.g. whole sensor initialization procedure can be stored in flash.
  - There are 3 message types:
    - transfer which contains `struct i2c_msg`
    - delay - sequence is delayed for milliseconds
    - pause - user provides function which is called when message is executed. Sequence is resumed using `i2c_async_resume()`.
  - `i2c_async` is a generic part and vendor specific driver implements only single transfer functionality (along with configure and getting async instance). This way specific driver is only focused on it's core functionality, all access management, semaphores, etc. is removed from the driver. It is in opposition to current state where each driver is implementing own mechanisms for access control.
  - `i2c_async` implements `z_i2c_async_sync_transfer` which is used to provide backward compatibility (it is called by `i2c_transfer`).

- Modified `i2c_nrfx_twim` driver to implement `i2c_single_transfer`

- Modified `hts221` sensor driver to show how to use i2c_async. This is temporary of course, just to show new way.

### How to control I2C using `i2c_async`?

Asynchronous access to i2c does not happen through i2c API because there can be multiple context accessing same i2c device. Access must be managed. `queued operation manager` is created to handle that generic task (see #23333) thus user is actually talking to `queued operation manager` API using handle to the manager associated with given i2c device. The manager works with `queued operations` object so i2c operation must contain queued operation structure. i2c operation consists of one or more `i2c_msg`. Messages build a sequence which is executed by the queue operation manager. Completion of the i2c operation is notified using `async notify` object (see #23229) which is part of the i2c operation. `async notify` standardizes methods of notification of completion of asynchronous operation.

Scheduling of asynchronous i2c is performed in following steps:
- create i2c operation (see `struct i2c_async_op` in `i2c_async.h`). Helper macros can be used to create it. Operation contains queued operation object and pointer to the sequence (pointer because sequence can be in flash and operation must be in rw memory).
- get `i2c_async` object to use. Use `i2c_get_async(dev)`.
- initialize `async_notify`. Use `i2c_async_get_notify(op)` to get async notify from the operation.
- submit operation `queue_operation_submit(mngr, op, PRI)`
- wait for notification about the completion

Example code:
```
/* create an operation with sequence */
I2C_ASYNC_OP_DEFINE(op, I2C_SENSOR_ADDRESS,
		I2C_ASYNC_MSG_TXRX(&reg_who_i_am, 1, &who_i_am, 1),
		I2C_ASYNC_MSG_PAUSE(who_i_am_validate),
		I2C_ASYNC_MSG_TX(reg_ctrl1, 2),
		I2C_ASYNC_MSG_DELAY(3),
		I2C_ASYNC_MSG_TXRX(&reg_conv, 1, conv_data, sizeof(conv_data))
	);

/* async instance */
struct i2c_async *i2c_async = i2c_get_async(i2c_dev);
struct async_notify *anotify = i2c_async_get_notify(&op);

/* get callback notification */
async_notify_init_callback(anotify, i2c_callback);

/* submit operation */
err = queued_operation_submit(i2c_async_get_qop_mgr(i2c_async), &op.op,
					QUEUED_OPERATION_PRIORITY_APPEND);
if (err < 0) {
	return err;
}
```

### Benefits
- operations can be submitted from interrupt context
- backward compatibility
- driver code focused on its functionality (single transfer)
- no need to extend i2c API with generic API (queued operation manager API used directly)
- similar approach can be applied to spi
- standardized asynchronous notification (async_notify)
- possibility to combine mulitple i2c operations into a sequence (e.g. init)
- possibility to pause a sequence (not added yet to i2c_async) which allows to combine operation like trigger sensor, wait for data rdy pin, read sensor into a single sequence.
- possibility to have single sensor implementation for spi or i2c bus with wrapper for creating messages (e.g. `SENSOR_BUS_MSG_TXRX()`, `SENSOR_BUS_OP_DEFINE`)

### misc

PR is based on unmerged work:
- `async_notify` #23229 
- `queued_operation` #23333
- `async_seq` #23075 

hts221 sensor changes commit will be removed from final PR

### CODE REVIEW

File: include/drivers/i2c.h [hunk 1]
```diff
@@ -212,6 +296,52 @@ static inline int z_impl_i2c_configure(struct device *dev, u32_t dev_config)
...
+ * @param callback Callback called on transfer completion.
+ * @param user_data User data passed to the callback.
+ *
+ * @retval Non-negative on successful transfer scheduling.
+ * @retval -EBUSY if device is currently performing a transfer.
+ * @retval -EINVAL if any parameter is invalid.
+ */
+static inline int i2c_single_transfer(struct device *dev,
+					const struct i2c_msg *msg, u16_t addr,
+					i2c_transfer_callback_t callback,
```
Comment (tbursztyka) [inline include/drivers/i2c.h]: I would avoid this. Primarily because it's going to be called in interrupt context (I know, gpio does it, but we agreed about that when it was designed to do so). In SPI we use struct k_poll_signal instead for notifying when a call is done (the only difference being that the call does not return -EBUSY but waits until it gets its turn but that's another story). Using such kernel object is I think better since it provides more flexibility and robustness as well as avoiding to run a cb in interrupt context.

File: include/drivers/i2c.h [hunk 2]
```diff
@@ -212,6 +296,52 @@ static inline int z_impl_i2c_configure(struct device *dev, u32_t dev_config)
...
+int z_i2c_async_sync_transfer(struct device *dev, struct i2c_msg *msgs,
+			    u8_t num_msgs, u16_t addr);
+
+/** @brief Get handle to the object for asynchronous transfers.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ *
+ * @return Asynchronous object or NULL if not supported.
+ */
+static inline struct i2c_async *i2c_get_async(struct device *dev)
```
Comment (tbursztyka) [inline include/drivers/i2c.h]: I see why you are doing this: it's easier then to manage async object memory space, as each and every driver would own it directly. But on the other hand, it is an object the driver will _never_ use by itself.
I think this needs to be reconsidered, finding a way to manager async objects directly into the async manager.

File: include/drivers/i2c.h [hunk 3]
```diff
@@ -121,6 +133,61 @@ struct i2c_msg {
...
+		.flags = I2C_MSG_WRITE | I2C_MSG_STOP \
+	}
+
+/** @brief Macro for initializing TX message
+ * @param txbuf TX buffer.
+ * @param txlen TX data length.
+ * @param delay Delay (in milliseconds) before starting TX if executed in
+ *		sequence.
+ */
+#define I2C_MSG_TX(txbuf, txlen, delay) \
```
Comment (tbursztyka) [inline include/drivers/i2c.h]: Could you have a separate patch for all the above? (Am I wrong if I say it could be used on i2c_transfer()? thus why I think it deserves its own patch.)

File: drivers/sensor/hts221/hts221.h [hunk 4]
```diff
@@ -29,9 +30,17 @@
 #define HTS221_REG_CONVERSION_START	0x30
 
 struct hts221_data {
+	struct i2c_async *i2c_async;
+	struct i2c_async_op op;
+	struct k_sem *tmp_sem;
 	struct device *i2c;
-	s16_t rh_sample;
-	s16_t t_sample;
+	union data {
```
Comment (tbursztyka) [inline drivers/sensor/hts221/hts221.h]: remove 'data'

**Review by tbursztyka — COMMENTED** [review]
The example is maybe not the best one (at least the init process: it looses the error check on chip id afaik)
And I did not realize this meant _that_ much change in user code to switch to async, and it will - at this stage - make it painful for maintainers to support both modes (do not forget that some users may not need such advanced feature). I think the most important part to improve is this then.

File: include/drivers/i2c.h (hunk 1 above)
Comment (pabigot) [reply include/drivers/i2c.h]: If this is supposed to be an asynchronous API, I'd really like to see it use `async_notify` rather than assume a callback is the user's desired interface for results.

**Comment by pabigot** [comment]
>   * Extending `i2c_msg` flags to contain milliseconds delay field which can be used within sequence to delay message execution.

I'm not a fan of packing up many features into a single API that would have be be independently supported in multiple drivers.  Why should a single-transaction API include a start delay?

Better to have a multi-operation description that steps through a sequence of operations, one of which could be a generic "delay" (for time, or for event) that is interpreted by an service-agnostic multi-operation sequencing engine, while service-specific operations are passed off to a common driver function like `process_operation`.

File: include/drivers/i2c.h (hunk 1 above)
Comment (tbursztyka) [reply include/drivers/i2c.h]: async_notify would mean such async function would be usable only within the async library. It may not be what all users want. They might want to use a simpler system, thus why using k_poll_signal might be more relevant.

Comment (pabigot) [reply include/drivers/i2c.h]: `async_notify` (from #23229) is simply an abstraction that describes how the caller wants to be notified that an operation completes.  It allows callback, poll (signal)l, or spin-wait notifications, so can be used from a variety of contexts with or without `POLL` support.

Comment (tbursztyka) [reply include/drivers/i2c.h]: Sure, however result and flags attribute seem useless (signal owns already a result attribute, and for the cb it could just take a result as parameter). Would be worth having an intermediate structure just for signal or cb then.

Comment (pabigot) [reply include/drivers/i2c.h]: I think it makes sense to discuss the `async_notify` capability in #23229 rather than here.  There are reasons why those fields are the way they are.

File: include/drivers/i2c.h (hunk 2 above)
Comment (nordic-krch) [reply include/drivers/i2c.h]: i agree. currently i had to do something to get this object. I already made a proposal of different approach: #22409

File: include/drivers/i2c.h (hunk 1 above)
Comment (nordic-krch) [reply include/drivers/i2c.h]: @tbursztyka 
> I would avoid this. Primarily because it's going to be called in interrupt context 

but that's the point. It is going to be called in the interrupt context and next transfer will also happen in the interrupt context. Operations are fully asynchronous. queue operation is scheduled from any context, then in transfer completion callback next one is taken from the sequence, if sequence is completed user is notified and queue is checked if there is another request pending and if yes next i2c transfer starts from interrupt context.

One thing to note is that `i2c_single_transfer` is not suppose to be used by the user, only by the `i2c_async` module. User can do that i.e. there is one sensor on the bus but then he gives up on access management thus scalability. Zephyr by default will use queue operation api (once/if ever all drivers switch to it).

@pabigot 
Regarding not using `async_notify` one of the reasons is the one mentioned above. It is not really intended to be used apart from i2c_async so it is tailored for that and adding async notify seems like overkill.

File: include/drivers/i2c.h (hunk 3 above)
Comment (nordic-krch) [reply include/drivers/i2c.h]: yes, except for delay parameter which was added to the flags. but you are right, i will extract it to separate patch.

**Comment by nordic-krch** [comment]
@tbursztyka 
> The example is maybe not the best one (at least the init process: it looses the error check on chip id afaik)

yes, it might be that chip id would need to be taken out, alternatively it can be checked after whole sequence. Another option that sequence manager is supporting is pausing. sequence can be configured so that after chip id read there is a pause callback. User checks chip id and aborts sequence if it is wrong or resumes. This is not yet implemented in i2c_async (supported by async_seq)

**Comment by nordic-krch** [comment]
@pabigot 
> Better to have a multi-operation description that steps through a sequence of operations, one of which could be a generic "delay"

I will look into that.

File: include/drivers/i2c.h (hunk 1 above)
Comment (tbursztyka) [reply include/drivers/i2c.h]: I still think you don't need to run any cb on interrupt context, for i2c that is. If reaaaaally needed, then have a in-between solution with a union on a cb and a k_poll_signal

Comment (nordic-krch) [reply include/drivers/i2c.h]: a union like that would be `async_notify` then.

Comment (tbursztyka) [reply include/drivers/i2c.h]: async_notify holds 2 useless attributes (result and flags), it can be another one.

Comment (pabigot) [reply include/drivers/i2c.h]: They are not useless.  See https://github.com/zephyrproject-rtos/zephyr/pull/23229#discussion_r391546701.  Please provide feedback there if you are unconvinced so we don't have to track discussions in two places.

Comment (nordic-krch) [reply include/drivers/i2c.h]: flags are needed to distinguish between callback or signal. and i advocate for moving result into union (it's needed for spinwait mode): https://github.com/zephyrproject-rtos/zephyr/pull/23229#discussion_r390947643

Comment (tbursztyka) [reply include/drivers/i2c.h]: oh yes, missed that entirely. Well, do you really need a interrupt context cb then? :)

Comment (nordic-krch) [reply include/drivers/i2c.h]: yes, i do. I don't want dedicated thread to talk to i2c. If i have signal then i need thread context to wait for the completion and i want to be able to do that from any context (interrupts).

**Comment by nordic-krch** [comment]
@tbursztyka 
> And I did not realize this meant that much change in user code to switch to async,

Note that legacy API will keep work and hts221 change was done only to showcase new api. I expect that asynchronous API will go in without any changes in sensors. It will serve for those who wants to use i2c asynchronously (custom driver). It's hard to tell if other vendors switch to that api. Since legacy and async can be mixed, sensor can have most parts common and for example two versions of read out call. However, this is too far into the future as that would require sensor API to be asynchronous.

**Review by pabigot — COMMENTED** [review]
From my perspective this PR moves too quickly.  It depends on two unmerged and mostly unreviewed PRs: 
* #24781 provides a way to submit multiple asynchronous requests to a single resource.  It's mine, and while I think it's good I've marked it DNM until we have evidence that it meets an actual need.
* #23075 proposes a way to synthesize a single uninterruptible complex operation from a sequence of primitive operations.

IMO because we're seeing three things at once here it's too hard to assess the I2C parts of this PR without first accepting the need for and implementation of the other two.  To the degree this PR demonstrates motivation for the other two enhancements in the context of I2C I think we should look at it in the wider context of multiple driver subsystems.

There are two existing subsystem driver APIs that provide synchronous and asynchronous interfaces for core operations: SPI has transceive, read, and write; ADC has read. (There's also one in clock, but it's prematurely stable as it doesn't meet Nordic's needs hence #24334).  These accept a single operation and a `struct k_poll_signal *`, and execute in another thread or interrupt context to reach the point where the operation completes.  (Attempts to submit a second asynchronous operation from the same or another thread **block** which is unacceptable and is the motivation for #24781.)

At this time the only asynchronous notification mechanism that can be used from user mode is signal, which is why those APIs use it.  I believe that @andrewboie is looking into the possibility of supporting pre-registered callbacks as will, which if it works out would open up the possibility of using the generic [asynchronous notification](https://docs.zephyrproject.org/latest/reference/misc/notify.html) API as a standard for the primitive single-operation async interface in a driver (subsystem).

Until that happens, no API that could be invoked from user mode should use `sys_notify` as its completion indicator.

So IMO the first step should be a standalone PR that provides async I2C capability using signal notification.  I would use `i2c_write_read_async()` as the basic operation, since it's a basis for most of the other operations (even if `i2c_write_read()` is currently implemented in terms of `i2c_transfer`), but `i2c_transfer` might be a better candidate.  I think there will be enough complexity in adding that one function.

Then we can extend the driver to be a queued operation service that manages asynchronous operations.  That would give us true asynchronous support for SPI, ADC, and I2C.

Then we can use a sequence (or transaction) manager to aggregate primitive operations to be invoked as a single transaction.

**Comment by nordic-krch** [comment]
@pabigot the point of this PR is to evaluate queued operation and sequence manager and show how i2c (spi) can benefit from them. It is also an attempt to fill the main gap I see in spi and i2c - lacking of isr callable functions. Is there a gap when it comes to threads/user space? Is there any issue for that, anyone requesting?

> IMO because we're seeing three things at once here it's too hard to assess the I2C parts of this PR 

I want to give short presentation on API meeting with overview of this approach.

**Comment by pabigot** [comment]
Notes from 2020-06-30 API telecon.  Edit/reply if I missed something.

TBD: Where the original slides can be obtained.  What's below are my notes.

### Presentation from @nordic-krch

Driver: be able to initiate I2C operations from interrupt context.
Requests to make more generic.  Presented overview:
* Queued operation manager, allows submission while device is busy (#24781)
* Sequence manager with generic actions, notifies when sequence complete (#23075)
* Queued sequence, shim to support sequence manager with operation manager (this PR)
* I2C API changes using these capabilities (this PR)

#### Sequence Manager:

* Sequence is immutable (pointer to array of actions)
* Operation mode:
  * manager provides processing function
  * action provides processing function
* Shared actions:
  * pause (invoke callback to allow sequence to resume)
  * delay (time-based pause)

Macros to:
* create the sequence `SYS_SEQ_DEFINE`
* create custom action `SYS_SEQ_CUSTOM_ACTION`

#### Queued sequence
* queued operation triggers sequence execution

#### I2C
* `i2c_single_transfer`
* `i2c_get_async(dev)` gets access to the queued operation manager
* `i2c_async` common implementation for actions

Actions:
* address set
* tx
* txv (buffer plus length)
* tx_rx
* txv_rx

#### Example using HTS221:
* set address, read who-am-i,
* pause verify result

#### Future:
* WAIT_FOR_PIN action
* Implement for SPI
* Combine I2C and SPI (if this makes sense)

JW: likes the script

### Questions

#### Agreement that interrupt-initiation of operations is important?

JF: doesn't seem necessary

KC: low power devices can make progress directly from interrupt context.  Also reduces RAM requirements for dedicated thread.

JF: most now use work queue.  Could use sensor work queue to decouple from impact on system work queue.

#### Complexity of sequence manager?

PB: The action sequence is getting close to describing a simple finite state machine.  Seems slippery slope: currently linear, looping within actions.  Will this add branching, control structures, ....?  Would that result be maintainable?

KC: Some branching support using return value of action to adjust selection of next action to execute.

### Summary

Awaiting further input on the specific questions above and the general concept.  Please comment here or in the specific component PRs.

**Comment by nordic-krch** [comment]
Presentation: https://docs.google.com/presentation/d/1C6C5Dka_jGIEnTqLROFAs1HTCkK3z_mL64ZH4TzLA-c/edit?usp=sharing

**Comment by pabigot** [comment]
API 2020-07-07: Moved from triage to in-progress pending further activity per https://github.com/zephyrproject-rtos/zephyr/pull/23371#issuecomment-651920879
